ipfs-cluster-service/migrate.go

```diff
@@ -0,0 +1,91 @@
+package main
+
+import (
+	"bufio"
+
+	"fmt"
+	"os"
+
+	"github.com/urfave/cli"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+	
+	ipfscluster "github.com/ipfs/ipfs-cluster"
+	"github.com/ipfs/ipfs-cluster/state/mapstate"
+	"github.com/ipfs/ipfs-cluster/consensus/raft"
+)
+
+
+func upgrade(c *cli.Context) error {
+	if c.NArg() < 1 || c.NArg() > 2 {
+		return fmt.Errorf("Usage: <BACKUP-FILE-PATH> [RAFT-DATA-DIR]")
+	}
+	//Load configs                                                             
+	cfg, clusterCfg, _, _, consensusCfg, _, _, _ := makeConfigs()
+	err := cfg.LoadJSONFromFile(configPath)
+	if err != nil {
+		return err
+	}
+	backupFilePath := c.Args().First()
+	var raftDataPath string
+	if c.NArg() == 1 {
+		raftDataPath = consensusCfg.DataFolder
+	} else {
+		raftDataPath = c.Args().Get(1)
+	}
+
+	//Migrate backup to new state                                                              
+	backup, err := os.Open(backupFilePath)
+	if err != nil {
+		return err
+	}
+
+	defer backup.Close()
+	r := bufio.NewReader(backup)
+	newState := mapstate.NewMapState()
+	err = newState.Restore(r)
+	if err != nil {
+		return err
+	}
+	//Record peers of cluster                                                                  
+	var peers []peer.ID
+	for _, m := range clusterCfg.Peers {
+		pid, _, err := ipfscluster.MultiaddrSplit(m)
+		if err != nil {
+			return err
+		}
+		peers = append(peers, pid)
+	}
+	peers = append(peers, clusterCfg.ID)
+	//Reset raft state to a snapshot of the new migrated state                                 
+	err = raft.SnapshotReset(newState, consensusCfg, raftDataPath, peers)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+
+func needsUpdate(cfg *ipfscluster.Config, cCfg *raft.Config) bool {
+	state := mapstate.NewMapState()
+	r, err := raft.ExistingStateReader(cCfg) // Note direct dependence on raft here
+	if err == nil { //err != nil no snapshots so skip check
```

Comment:
if validSnap == false && err == nil -> no snapshots / nothing to do
if validSnap == true && err == nil -> do GetVersion check and maybe tell user migration is needed
if validSnap == true && err != nil -> could not deserialize state from snapshot, migration very probably needed or snapshot corrupted
if validSnap == false && err != nil -> some error before trying to read snapshots